### PR TITLE
PIM-5110: Add product export configuration to customize product media export policy

### PR DESCRIPTION
**Pull request by ClementGautier**
## Add export parameter to export product media or not

screenshot from 2016-07-07 12 11 49

## DoD

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | N
| Added Behats                      | Y
| Changelog updated                 | Y
| Review and 2 GTM                  | Y
| Micro Demo to the PO (Story only) |
| Migration script                  | N
| Tech Doc                          | N

### CODE REVIEW

**Comment by ClementGautier** [comment]
GTM (The only that could bother me is that you test again things that should be already covered, bu I prefer too much tests than a missing test)

File: src/Pim/Component/Connector/spec/Writer/File/CsvProductWriterSpec.php [hunk 1]
```diff
@@ -109,4 +110,38 @@ function it_prepares_the_export(
             '/tmp/export' => 'export'
         ]);
     }
+
+    function it_does_not_copy_media_if_parameters_with_media_is_falsy(
```
Comment (skeleton) [inline src/Pim/Component/Connector/spec/Writer/File/CsvProductWriterSpec.php]: is_fals**e**

File: src/Pim/Component/Connector/spec/Writer/File/XlsxProductWriterSpec.php [hunk 2]
```diff
@@ -135,6 +136,40 @@ function it_writes_the_xlsx_file(
         $this->flush();
     }
 
+    function it_does_not_copy_media_if_parameters_with_media_is_falsy(
```
Comment (skeleton) [inline src/Pim/Component/Connector/spec/Writer/File/XlsxProductWriterSpec.php]: same

Comment (skeleton) [inline src/Pim/Component/Connector/spec/Writer/File/XlsxProductWriterSpec.php]: no spec with `with_media` as `true` ?

File: src/Pim/Bundle/ImportExportBundle/JobParameters/FormConfigurationProvider/ProductCsvExport.php [hunk 3]
```diff
@@ -173,10 +190,8 @@ public function getFormConfiguration(JobInstance $jobInstance)
                 ]
             ],
         ];
-        
-        $formOptions = array_merge($formOptions, $this->simpleCsvExport->getFormConfiguration($jobInstance));
 
-        return $formOptions;
+        return array_merge($productFormOptions, $csvFormOptions);
```
Comment (skeleton) [inline src/Pim/Bundle/ImportExportBundle/JobParameters/FormConfigurationProvider/ProductCsvExport.php]: why two array_merge ?

File: features/Pim/Behat/Context/Domain/Spread/ExportProfilesContext.php [hunk 4]
```diff
@@ -138,6 +138,31 @@ public function theNameOfTheExportedFileOfShouldBe($code, $path)
...
+    }
+
+    /**
+     * Check if files $shouldBe or not in the export directory of the job with the given $code
+     *
+     * @param string    $code     Code of the job instance
+     * @param bool      $shouldBe true if the files should be in the directory, false otherwise
+     * @param TableNode $table    Files to check
+     */
+    protected function checkExportDirectoryFiles($code, $shouldBe, TableNode $table)
```
Comment (skeleton) [inline features/Pim/Behat/Context/Domain/Spread/ExportProfilesContext.php]: $shouldBeInDirectory ?

File: src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php [hunk 5]
```diff
@@ -67,6 +62,16 @@ public function write(array $items)
         $parameters = $this->stepExecution->getJobParameters();
         $withHeader = $parameters->get('withHeader');
         $this->flatRowBuffer->write($products, $withHeader);
+
+        $exportDirectory = dirname($this->getPath());
```
Comment (skeleton) [inline src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php]: create directory after you test if `with_media` is true

**Comment by skeleton** [comment]
gtm once fixed

File: src/Pim/Bundle/ImportExportBundle/JobParameters/FormConfigurationProvider/ProductCsvExport.php (hunk 3 above)
Comment (grena) [inline src/Pim/Bundle/ImportExportBundle/JobParameters/FormConfigurationProvider/ProductCsvExport.php]: @skeleton It's to ease readability, to see that we merge the base csv form options with the dedicated product form options. But if you think it's bad, I can keep one array_merge.

File: src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php (hunk 5 above)
Comment (grena) [inline src/Pim/Component/Connector/Writer/File/XlsxProductWriter.php]: It was, but it doesn't work as the XLSX files can't be moved in this directory if it doesn't exist.
The test is done after in the CSV writer because it manipulates files directly, not the XLSX writer, seen with @jjanvier

File: src/Pim/Bundle/BaseConnectorBundle/Archiver/ArchivableFileWriterArchiver.php [hunk 6]
```diff
@@ -107,7 +107,7 @@ protected function getZipFilesystem(JobExecution $jobExecution, $zipName)
      */
     protected function isWriterUsable(ItemWriterInterface $writer)
     {
-        return $writer instanceof ArchivableWriterInterface && count($writer->getWrittenFiles()) > 1;
+        return $writer instanceof ArchivableWriterInterface;
```
Comment (jjanvier) [inline src/Pim/Bundle/BaseConnectorBundle/Archiver/ArchivableFileWriterArchiver.php]: why did you remove it?

**Comment by jjanvier** [comment]
imho, `archive` is not a good naming:
* because I'm not sure POs use it (maybe they use `zip`)
* because we use this term for the files that are put in `app/archives/` (ie: every import/export file)

**Comment by jjanvier** [comment]
and more globally, I don't understand why this story has been taken now (if the archiving system is redesigned, everything here will be broken... and done twice... by 2 different teams...)

but it's an organizational problem

File: src/Pim/Bundle/BaseConnectorBundle/Archiver/ArchivableFileWriterArchiver.php (hunk 6 above)
Comment (grena) [inline src/Pim/Bundle/BaseConnectorBundle/Archiver/ArchivableFileWriterArchiver.php]: If the `archived` param is `true`, we have to archive, even if there is only one file

**Comment by grena** [comment]
@jjanvier For the naming, maybe "compress"?
I'd prefer avoid `zip` as it's only a format. But I've got your point!

**Comment by jjanvier** [comment]
@grena the naming is determined by the business (maybe with the help of developers)
so you have to ask the POs

**Comment by grena** [comment]
Seen with POs, we keep "archived" :water_buffalo:

**Comment by grena** [comment]
CI is green. I'm squashing some un-relevant commits and I merge :)
